### ISSUE TITLE
feature suggestion for `is` compare between different types

### ISSUE THREAD

**Issue by orSolocate** [PyCQA/pylint#5763]
### Current problem

Comparing two different types using the `is` operator is always `False`.

For example for this code:

```
A = 5
STR1 = ‘5’
if A is STR1:
	pass
```

`A` is of type `int` and `STR1` is of type `str`.
`is` compare between those types is always `False`.

The user should be alert about it, and consider modifying his code.

As far as I checked there is no checker that checks that..

### Desired solution

I suggest a new checker (or extension of an existing one) that raises a message for this case.

message name suggestion: `is-different-types` or `different-types-is-expression`

message type: `Warning`

message content suggestion:

`An 'is' comparison between different types always fails, consider using casting`

### Additional context

The check should be general to any comparison/expression, not necessarily an `if` comparison.
I think we should raise a message even for cases like:

`condition = '5' is 4`

I thought about extending the checker to a `==` operator, but I am not really fond of the idea and I will explain why.

1. The `__eq__` of Python classes can be modified to return `True` for different comparison values.

2. There is a (stupid) case where `==` actually returns `True` for different types. it happens for `int` and `float`, e.g. `1.0==1`.
[note that the statement `1.0 is 1` returns `False`, as `1.0` is of type `float` and `1` is of type `int`]
[also note that this rare case only happens for small numbers]

**Comment by orSolocate** [comment]
Edit: actually I don't think casting should be the suggestion here. It would still be `False` after casting...

**Comment by orSolocate** [comment]
I can implement it if there is value in that for Pylint.
@Pierre-Sassoulas  or anyone else like the idea?

**Comment by Pierre-Sassoulas** [comment]
I'm hesitant about this one because it seems like a lot of effort with potentially a lot of false positive where the user knew what they're doing. There's also a lot of case that are already handled by thing like [flake8's E711](https://www.flake8rules.com/rules/E711.html) or pylint's literal-comparison, which is why I added the discussion label to see what others are thinking about it.

**Comment by DanielNoord** [comment]
I'm having similar though as @Pierre-Sassoulas. I wonder how often this would catch actual bad code compared to the number of edge use cases that could turn into false positives.

**Comment by orSolocate** [comment]
could you explain why you think there will be false positives with an example?

The implementation can be restricted only to `if` visits. it can be reduced to work only on either built-in types or inferable names. otherwise ignored. This will increase false negatives and decrease false positives.

**Comment by DanielNoord** [comment]
I can't think of any right now, however, I wonder how many actual messages will be raised. How often do people use this pattern when they shouldn't? And how often do people use this when they should?
I'm not totally convinced of the right balance between 1) catching actual errors and 2) the risk of emitting a number of false positives.

**Comment by Pierre-Sassoulas** [comment]
I'm adding  the need decision label and PR / investigation label as seeing the effect on the primer would add valuable input to the decision.

**Comment by orSolocate** [comment]
I will try to check it when I have the time (keep this issue opened)
